app/route53.py

```diff
@@ -0,0 +1,248 @@
+
+    fsd = _derive_fully_specified_domain(url)
+    zone_id = None
+    for z in zones['HostedZones']:
+        if fsd == z['Name']:
+            zone_id = z['Id']
+
+    if zone_id is None:
+        msg = 'Could not find hosted zone for fully specified ' \
+              'domain: {0}'.format(fsd)
+        log.error(msg)
+        log.error(pformat(zones))
+        raise Route53Error(msg)
+
+    log.info('Using HostedZoneId: {0}'.format(zone_id))
+    return zone_id
+
+
+def _find_cname_record(client, zone_id, cname_domain):
+    """Find an existing record for the `cname_domain`, or `None` if one does
+    not exist.
+
+    Parameters
+    ----------
+    client :
+        Boto3 Route 53 client.
+    zone_id : str
+        The Hosted Zone ID for the CNAME.
+    cname_domain : str
+        The CNAME domain. See
+        http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/ResourceRecordTypes.html#CNAMEFormat
+
+    Returns
+    -------
+    record : dict
+        The ``ResourceRecordSet`` with a ``Name`` corresponding to the
+        `cname_domain`. E.g.,
+
+        .. code-block:: python
+
+           [{'Name': 'www.mydomain.org.',
+             'ResourceRecords': [{'Value': 'origin_domain.com'}],
+             'TTL': 14400,
+             'Type': 'CNAME'}]
+
+        `None` is return is not ResourceRecordSet matching `cname_domain` is
+        found.
+    """
+    # turns out boto3 doesn't need StardRecordName in lexicographic order
+    # despite their docs.
+    # name = _lexicographic_order_domain(cname_url)
+
+    if not cname_domain.endswith('.'):
+        cname_domain = cname_domain + '.'
+
+    r = client.list_resource_record_sets(
+        HostedZoneId=zone_id,
+        StartRecordName=cname_domain,
+        StartRecordType='CNAME'
+    )
+    if r['ResponseMetadata']['HTTPStatusCode'] != 200:
+        msg = 'list_resource_record_sets failed:\n' + pformat(r)
+        log.error(msg)
+        raise Route53Error(msg)
+    for record in r['ResourceRecordSets']:
+        if record['Name'] == cname_domain:
+            log.info(pformat(record))
+            return record
+
+    log.info('No existing CNAME record found for {0}'.format(cname_domain))
+    return None
+
+
+def _upsert_cname_record(client, zone_id, cname_domain, origin_domain):
+    """Upsert a CNAME record of `cname_domain` that points to `origin_domain`.
+    """
+    change = {
+        'Action': 'UPSERT',
+        'ResourceRecordSet': {
+            'Name': cname_domain,
+            'Type': 'CNAME',
+            'TTL': 900,
+            'ResourceRecords': [
+                {
+                    'Value': origin_domain
+                },
+            ],
+        }
+    }
+    change_batch = {
+        'Comment': 'Upsert {0} -> {1}'.format(cname_domain, origin_domain),
+        'Changes': [change]}
+    log.info(pformat(change_batch))
+
+    r = client.change_resource_record_sets(
+        HostedZoneId=zone_id,
+        ChangeBatch=change_batch)
```

Comment:
I don't know if boto handles this for you but CNAME entries should be qualified from the root (`.`).  Eg.,

```
$ dig ts-ci.lsst.codes +noall +answer

; <<>> DiG 9.10.3-P3-RedHat-9.10.3-10.P3.fc23 <<>> ts-ci.lsst.codes +noall +answer
;; global options: +cmd
ts-ci.lsst.codes.	278	IN	CNAME	ci.ts.lsst.codes.
ci.ts.lsst.codes.	278	IN	A	52.36.85.108
```